TheJumpCloud/jumpcloud-ADMU
# Sa 2995 mtp support

## Issues
* [SA-2995](https://jumpcloud.atlassian.net/browse/SA-2995) - MTP Support for ADMU

## What does this solve?

Administrators with MTP admin keys could not identify users to bind in previous versions of the ADMU if they were using their MTP API Key. This release addresses this limitation by specifically prompting admins to select a MTP Org if a MTP Key is detected.

## Is there anything particularly tricky?

Yes, many functions were changed in this module update. A MTP key is required to test this functionality. 

## How should this be tested?

The GUI App must be validated manually, test cases are written for the command line version of the tool

[Test Case: GUI Migration (MTP)]

- With a Windows VM not bound to any tenant
- Install JumpCloud on the VM and add to an MTP tenant 
- Run the JumpCloud GUI from this branch
- Select AutoBindJC user and add an **MTP API Key**
- Select an Org
- type a user to perform a username search
- migrate a user with the autobind JC username box checked
- The user should be matched with a user in the MTP tenant, migrated and bound to the device after ADMU

[Test Case: CLI Migration (MTP)]

- With a Windows VM not bound to any tenant
- locally import "Start-Migration.ps1"
- Run the ADMU command line version on a AD user ex:
- `Start-Migration -JumpCloudUserName "james.lowery" -SelectedUserName "S-1-12-1-64447351-1339341875-2832918416-977843340" -TempPassword T#st123456789 -InstallJCAgent $true -JumpCloudConnectKey "MTP_Tenant_Connect_Key" -AutobindJCUser $true -JumpCloudAPIKey "MTP_API_KEY" -JumpCloudOrgID "MTP_ORG_ID"`
- The agent should be installed, the user should be matched with a user in the MTP tenant, migrated and bound to the device after ADMU

## Screenshots

New GUI displays OrgName/ OrgID after an API Key is validated:
![Screen Shot 2022-11-16 at 8 46 53 AM](https://user-images.githubusercontent.com/54448601/202229074-846e4155-cabd-4afc-9894-d4c8e1b840c7.png)

When an MPT Key is validated a GUI prompt is presented, admins can select one org from a dropdown list. Org IDs are presented because Org Names are not unique:
![Screen Shot 2022-11-16 at 8 47 19 AM](https://user-images.githubusercontent.com/54448601/202229299-6497d1fa-88a4-462f-ae20-690b891afa20.png)

Changed behavior:
When the AutoBind JumpCloud User option is selected. A validation step now occurs and the API Key is checked before migration begins.

If the Key is an MTP Key and no Organization ID is provided, the script will error and exit
![Screen Shot 2022-11-16 at 9 44 31 AM](https://user-images.githubusercontent.com/54448601/202243263-6c5dadf6-29f9-4387-bf86-ed8ca4147188.png)

If the ORG ID is provided, the script will validate that the Key and ORG ID match. This behavior is the same for a non-MTP Admin key. As such there should be no need for Non-MTP admins to edit their scripts and pass in an ORG ID. MTP Admins will need to do this
![Screen Shot 2022-11-16 at 9 48 09 AM](https://user-images.githubusercontent.com/54448601/202243298-90c7fe01-907d-431a-971a-772081c7208c.png)
